### ISSUE TITLE
Labels for relay types

### ISSUE THREAD

**Issue by vitorpamplona** [nostr-protocol/nips#1282]
We need to find good labels for each relay type. Apps need to request users to put a relay of a given type in certain fields and relays need help to advertise which type they are made for. 

- Public Outbox Relays: Relays that store all the public content of a user in a way that anyone out there can download them. This is a requirement to find posts on NIP-65 clients. 

- Public Inbox Relays: Relays that accept any event from anyone as long as they tag one of the subscribers or the public in general. Anyone can download these events. Great for Notifications.

- Private Inbox: Relays that accept any event from anyone as long as they tag one of the subscribers or the public in general, however only the tagged individual can download their tagged events. In general, the public cannot download events. Great for DMs and RPC-like calls: NSec Bunker, Nostr Wallet Connect and DVM relays. 

- Private Storage: Relays that accept events from an author and make sure only the author can download them. This is great for Private Drafts or other private content the author is not sharing with anyone, like settings private pet names, follow lists, etc. 

- Search Relays: Relays that implement NIP-50 and can help find content. 

- Purple page Relay: Relays that only track kind 0 and kind 10002

- Community Relays: Relays whose write access is closed to members or a NIP-29 group or NIP-71 community. 

Are there any other relay types out there? 

| Type      | Who can Read | Who can Write |
| ----------- | ----------- | ----------- |
| Public Outbox      | Everyone       | Authors |
| Public Inbox   | Everyone        | Everyone (may need to tag a white-listed user) |
| Private Storage   | Authors        | Authors |
| Private Inbox   | Authors       | Everyone (may need to tag a white-listed user) |
| Search   | Everyone (or white-listed users if closed)       | Nobody (this relay is generally read only) |
| Purple Pages   | Everyone (or white-listed users if closed)       | Nobody (this relay is generally read only) |

@mikedilger

**Comment by mikedilger** [comment]
I had not thought of all of those since half of those are not things I intend to support. But this list makes sense.

I was calling purple page relays "Discovery" relays.

**Comment by lingmann** [comment]
Should there be a call-out for "caching" style relays similar to what Nostrudel is doing with nostr-rs-relay or nostr-relay-tray? If I'm understanding the distinctions above, the closest would be "Community" or "Local Database Relays".

I could see this becoming a pattern where one or more users run a local high-performance cache relay which they connect up to multiple clients, but only whitelisted users can read and write events (NIP-42 style or perhaps IP restricted, but any event is accepted).

**Comment by staab** [comment]
That list seems pretty complete to me, I agree it would be great to have all this as a first-class thing. For the "purple pages" relays, I prefer to call those "indexers", because they're basically a key-value store of pubkey -> profile/relays. In addition to kind 0 and 10002, they should probably include the blossom server selections document, as well as any other kinds that come out of this discussion intended to track relay selections. It might also be useful to include nip 89 handlers and lists, although that's probably overkill if you know the pubkey's relay selections. But if that seems useful, it could be worthwhile to add some sort of `indexed_kinds` key to NIP 11.

**Comment by vitorpamplona** [comment]
Both Discovery and Index are good words. Maybe Discovery is too generic.

**Comment by jeremyd** [comment]
Some thoughts:
I am tempted to describe these as relay features or usages, not 'types' as a relay can function as more than one of these categories..   As you start configuring relays to do certain things you realize "there is no relay types".  There are only usages from a client's perspective.

I also think clients should be able to understand more about a relay's capability before attempting to use the outbox model.  For example, I want to use mostly paid relays.  I know that most of my friends use the WINE relay and even though the description for "public inbox" says anyone should be able to write, this is un-true as I know a majority of the replies I want to see will come from WINE.  I may want to filter this further showing only replies from paid relays vs. replies from free relays (spam).

I think there may need to be levels of granularity, and also usage.  I have always thought the amethyst config screen is nice because each icon next to a relay entry represents a usage and you can toggle them on or off.  Perhaps something more along these lines would make sense instead of categorization by 'type'.

**Comment by jeremyd** [comment]
So to summarize, I would potentially do it more like gossip's relay config, where for each relay you configure it's usages and ALSO it's rank.  Then I would have the client be able to filter by rank...

**Comment by vitorpamplona** [comment]
Features work for us but not for users. We need to be able to say: "Find a relay that is made for Private Inbox and put the URL here". Then they need to search for relays that can do "Private Inbox". And from there they can evaluate the other details the relay providers, but there must be a way to reduce the options down.

**Comment by frbitten** [comment]
I imagine another type of relay or service for relays that I don't think is described in any NIP yet.

A relay that suggests events based on my preferences. It would be a relay for feed suggestions. Today, clients basically download everything from all relays, generating a lot of noise from irrelevant posts.

I believe that a relay that suggests events based on my preferences would bring NOSTR closer to the interactivity of large social networks and would also be a good weapon against spam and unwanted topics.

The competition between relays to provide a better service could evolve this functionality quickly.

**Comment by Yonle** [comment]
For the `Proxy`, Uploading is `Everyone (or white-listed users)` as it also does rebroadcasting events to other relays.

**Comment by dskvr** [comment]
Would be very interested in this for #230.

**Comment by jeremyd** [comment]
Here's an attempted re-wording for "Public Inbox Relays"

---
These relays are where your client looks for replies, comments, likes and zaps on your posts.  These relays can be paid or free relays, keep this in mind as it can limit the notifications you receive if the relay limits access.  For example if you are being attacked by comment spam, using paid relays here can be useful.
---

As a side note in relation to this above, I think it would be extremely useful to have filters on these notifications, similar to how there is a filter in amethyst for notifications "from follows" and "global".  Relays could have a rank, and you could filter by rank.  IE: set your paid relays as rank 1, and free relays rank 2.  Then you can filter just for rank 1, and later if you want more notifications and your attackers go away, you can expand to ranks 1 and 2.

Another attempted re-wording for DM Inbox Relays:
---
Insert between 1-3 relays to serve as your private inbox.  Others will use these relays to send DMs to you.  If you want to receive DMs from the open network then these relays must accept messages from anyone.  However if you only want verified, paid, or WoT users to be able to send you DMs then these relays can limit access to only these users.
Good options are: ...
---

General Relays:
I'd like to see this description expanded to what amethyst uses these for.  I'm unclear exactly when they come into play.

**Comment by vitorpamplona** [comment]
> Then you can filter just for rank 1, and later if you want more notifications and your attackers go away, you can expand to ranks 1 and 2.

Interesting idea.

I changed the description for public inbox. 

The private inbox however, can't do what you describe. The author of NIP-17 DMs is random. The relay choice cannot protect users against spam in DM (or gift wraps in general). The client must take that responsibility.

> General Relays

That is only a fallback for everything else in Clients that are migrating from Kind3 relay lists.

**Comment by jeremyd** [comment]
> > Then you can filter just for rank 1, and later if you want more notifications and your attackers go away, you can expand to ranks 1 and 2.
> 
> Interesting idea.
> 
> I changed the description for public inbox.
> 
> The private inbox however, can't do what you describe. The author of NIP-17 DMs is random. The relay choice cannot protect users against spam in DM (or gift wraps in general). The client must take that responsibility.

The author of the event is random yes.  However, with NIP42 auth, the authenticating user does not have to be random (IE, you're not doing an auth request with the random newly generated key are you?).  Using AUTH means, if the relay has an ACL and NIP42-AUTH enabled, the client can establish the connection with this and publish with as many other keys as they like.  This will cut back on un-solicited DMs as scammers will not be on the ACL as frequently as on a wide open relay.

> 
> > General Relays
> 
> That is only a fallback for everything else in Clients that are migrating from Kind3 relay lists.

**Comment by vitorpamplona** [comment]
Suggested two new relay lists for Proxy and Broadcasters (Publishers?) so that they don't end up mixed with other relay lists. https://github.com/nostr-protocol/nips/pull/1303

**Comment by jeremyd** [comment]
> Suggested two new relay lists for Proxy and Broadcasters (Publishers?) so that they don't end up mixed with other relay lists. #1303

I think it'd be nice to have a relay type that is 'hidden' IE, not advertised in your relay list and still used by the client.

**Comment by jeremyd** [comment]
> > General Relays
> 
> That is only a fallback for everything else in Clients that are migrating from Kind3 relay lists.

Sorry to clog the thread, maybe this is off-topic.  But, I still would like to know exactly what the list is used for.  Is amethyst still publishing kind3 relay lists with this? Is that *all it is doing?  Are these relays connected to when notes aren't found?  What happens if I remove them all?  Is the global icon going to be depricated?  Will there be any way to selectively create a global feed in the new versions?

**Comment by vitorpamplona** [comment]
> Is Amethyst still publishing kind3 relay lists with this?

Yes. Kind3 lists operate just like they were before NIP-65. Most new events will be sent to that list. The only ones that were already migrated are Drafts and NIP-17 DMs. 

> Are these relays connected to when notes aren't found? 

They are still connected at all times. So they can bring in new chats, products or other events that might not be on the NIP-65 lists. 

> What happens if I remove them all?

The app probably won't work very well at this point. Many sections of the app still rely on Kind 3 lists. 

> Is the global icon going to be depricated? Will there be any way to selectively create a global feed in the new versions?

We are likely to migrate to NIP-51 Relay sets (30002) for Global. Users will switch between relay lists from the Top bar when browsing global.

This might take a while though.

**Comment by cameri** [comment]
So far I think this classification is pretty good. I think we are missing Bridges to other networks like the ones built by @alexgleason.

**Comment by vitorpamplona** [comment]
Hey @nostr-wine, is filter.nostr.wine a proxy or just a broadcaster?

**Comment by nostr-wine** [comment]
Both! We read from and broadcast to a list of public relays.

https://docs.nostr.wine/filter/readme

**Comment by jeremyd** [comment]
This list is looking great.  I would like to use these types as tags in the NIP11 (relay info via http) spec, and also in NIP66 (relay discovery via events).  Once this is finalized can we have tag names as part of the standard.  We could use the "Type" mentioned here with dashes instead of spaces.

**Comment by vitorpamplona** [comment]
> This list is looking great. I would like to use these types as tags in the NIP11 (relay info via http) spec, and also in NIP66 (relay discovery via events). Once this is finalized can we have tag names as part of the standard. We could use the "Type" mentioned here with dashes instead of spaces.

Feel free to kickstart that. The list here will constantly change. As long as we call the same thing between relays and clients, we should be good.

**Comment by frbitten** [comment]
A relay that would have several functionalities. How would it identify these functionalities to the client?

I think it would be impractical to rely solely on the user to enter the address in the correct option.

In NIP-11, it would be possible to identify the types supported by the relay.

From the relay's point of view, would it be interesting for each type to have a different URI?

Something like:
- wss:111.111.11.1/blob for the Blob type
- wss:111.111.11.1/search for the search type.

**Comment by vitorpamplona** [comment]
> A relay that would have several functionalities. How would it identify these functionalities to the client?

For the sake of sales, marketing, and user experience, I would break it into many products, one for each thing. Even if all of them just map it back to one main server. The goal is just to simplify the life for users who have to choose their providers for each category. 

> I think it would be impractical to rely solely on the user to enter the address in the correct option.

Later on, it will be nice to have a new event kind to declare the relay service to the network with tags for each service type. Then apps just need to pull all self-declarations, filter by tag and offer those to the users.

**Comment by frbitten** [comment]
Would a dedicated relay for NIP-78 be classified as "Private Storage" or "Blob Relays"? Would a specific classification for this NIP make sense?

**Comment by staab** [comment]
> Would a dedicated relay for NIP-78 be classified as "Private Storage" or "Blob Relays"? Would a specific classification for this NIP make sense?

NIP 78 is use-case-agnostic, which means it's impossible to categorize. I'm not sure what to do about that.

**Comment by frbitten** [comment]
What if NIP-78 is for both? I don't know how confusing it would be for users if a nip could be in more than one type of relay.

**Comment by sondreb** [comment]
I think this thread is a great example of how Nostr is going to die soon (hyperbole, _I'm partially joking_).

There is no way Nostr will scale with this complexity, neither in terms of physical infrastructure and especially not in terms of requirements for Nostr client development, but finally also it will fundamentally kill adoption of Nostr due to complexity and bad UX.

We don't need these classifications nor having different relays for different purposes for our users.

My suggestion is to simply rely on kind:10002 (ignoring READ/WRITE) and jump across to the author's relays whenever you need to retrieve events such as kind:0 from those you have replied to you. This basic relay list is what should be understood as the user's "home". It is where all their content is stored, not across multiple relays implementing different protocols and capabilities.

**Comment by jeremyd** [comment]
> There is no way Nostr will scale with this complexity, neither in terms of physical infrastructure and especially not in terms of requirements for Nostr client development, but finally also it will fundamentally kill adoption of Nostr due to complexity and bad UX.
> 
> We don't need these classifications nor having different relays for different purposes for our users.

There are a few relay choices a user *must make if they care about a certain aspect, or "micro app" on nostr.  One of them is DMs.  If you want to do DMs currently on nostr, you will want to use kind 10050 relay lists.  Any use case similar to this, will do the same either with a specific relay list or relay hints etc.  I think the complication comes when you try to be an "Everything app".

> 
> My suggestion is to simply rely on kind:10002 (ignoring READ/WRITE) and jump across to the author's relays whenever you need to retrieve events such as kind:0 from those you have replied to you. This basic relay list is what should be understood as the user's "home". It is where all their content is stored, not across multiple relays implementing different protocols and capabilities.

basic relay list, yep, kind 10002 (are you saying get away from kind 3? I agree).  I generally have the same relays in read/write, (when you don't specify it's both).  You may want to have a read relay, if you are reading from a paid to post relay, but not posting.  Or you can just send there and the relay rejects it (not that great, but common).

**Comment by sondreb** [comment]
> basic relay list, yep, kind 10002 (are you saying get away from kind 3? I agree). I generally have the same relays in read/write, (when you don't specify it's both). You may want to have a read relay, if you are reading from a paid to post relay, but not posting. Or you can just send there and the relay rejects it (not that great, but common).

Absolutely, the main issue with kind 3 is that it's a large event and was never intended to hold the relay list, it was a hack all of us did in the beginning.

By separating relay list into its own event, we can utilize this file for a very scalable and decentralized network.

Relay List events (kind:10002) is the "Mainline DHT of Nostr" or "DID DOcument of Nostr".

**Comment by vitorpamplona** [comment]
I strongly reject the idea of putting all types of relays into a single list. 

Relay types will always exist because anyone can do whatever they want on Nostr, including build custom relay types to best match the needs of their client or use case. There is no way around this. This is an inherent feature, not a bug, and attempting to contain it is futile. We, as a group, cannot do anything about it. The proliferation of relay types is unavoidable and will grow massively over time. 

Those relays are not the same. They are generally not compatible with one another. If you use them incorrectly, some will overload your app with data you didn't ask for and don't need, some will never return anything and just waste mobile data, others will duplicate calls and double the mobile data usage, some will send records that were supposed to be for private (company, my own storage) relays into public ones, share health data in non-HIPAA-compliant relays, etc. 

That means that if we put all relay types in a single kind, we overload that kind and require every client to deal with or filter out every different relay out there. It unnecessarily creates a really dumb n*n problem. By separating them into their own lists, clients can decide when to support a specific kind, as they already do with NIPs, and thus fully ignore the majority of the relay sets out there that are not needed for their app. 

Multiple lists are just basic architectural hygiene.

**Comment by sondreb** [comment]
> I strongly reject the idea of putting all types of relays into a single list.
> 
> Relay types will always exist because anyone can do whatever they want on Nostr, including build custom relay types to best match the needs of their client or use case. There is no way around this. This is an inherent feature, not a bug, and attempting to contain it is futile. We, as a group, cannot do anything about it. The proliferation of relay types is unavoidable and will grow massively over time.
> 
> Those relays are not the same. They are generally not compatible with one another. If you use them incorrectly, some will overload your app with data you didn't ask for and don't need, some will never return anything and just waste mobile data, others will duplicate calls and double the mobile data usage, some will send records that were supposed to be for private (company, my own storage) relays into public ones, share health data in non-HIPAA-compliant relays, etc.
> 
> That means that if we put all relay types in a single kind, we overload that kind and require every client to deal with or filter out every different relay out there. It unnecessarily creates a really dumb n*n problem (every client needs to know every relay type). By separating them into their own lists, clients can decide when to support a specific kind, as they already do with NIPs, and thus fully ignore the majority of the relay sets out there that are not needed for their app.
> 
> Multiple lists are just basic architectural hygiene.

There are no precedence or examples of other software or architecture that does something like this, is there? IRC did not, HTTP did not. Client-Server software doesn't do this. Solid did not, Web5 did not.

Then, Nostr needs this, or it will fail?

If you have some piece of custom software that does something special on a Nostr relay, then that can easily be handled in your software, right? If I'm doing stream logging of IoT data from my equipment and delivering it to our partners, I'll announce our relays to the world where to grab that data.

Don't need to do much to "contain it", it's just a matter of interpretation of the NIPs and decision on what NIPs to adopt.

`Calle` as an example, have 14 relays in his Relay List. That's absolutely pointless and wasteful, it demonstrates the issue at hand here and how there are no clear way to scale Nostr. He got 50,000 followers. Only one of his relays is marked `read`, the others have no markers, meaning that every single reply to `Calle` is being written to 14 different relays.

Yourself have 4 relays, two with read/write, one write, one read. You have indicated that your content should be read from a paid relay, but that we should write to some free ones, and you host your own private one as well? So, 3 of 4 is where we should publish replies to you, and 3 of 4 you want us to read your events from. How does this give the network scalability or simplicity? It adds complexity for very little gained value. My client will attempt to publish to all 4 of your relays.

Users should use same relays for all their data, and if they want improved services, they could pay for a premium relay. That premium relay should obviously allow delivery of events to their paying customers.

The mobile data usage is already out of hand, it won't get better continue down the similar path.

Users need 2-4 relays each, not more. As a matter of fact, for user's who have more, I won't connect to them for read, just 2 or 4 at a time. Popular users can (and should) pay for their own popularity and clients can (and should) perform random round robin connections towards the relays, 1 or 2 at a time. That's how large accounts can scale, if I have a lot of people interested in my data, I musts provide a larger set of relays that my followers can connect to, to get my data.

I will at least make an attempt at implementing my own suggestions and we can see how that goes, if it is viable manner in which Nostr will scale and improve UX. It's the only thing that matters, showing it can be done.

Thanks!

**Comment by vitorpamplona** [comment]
> Then, Nostr needs this, or it will fail?

It's not a matter of needs. It's a matter of "those relay types exist and no one can remove them from existing".

> Calle as an example, have 14 relays in his Relay List. That's absolutely pointless and wasteful

Correct. But he is following what you preach... putting everything in the NIP-65 kind. That's what some clients also preach (Primal, Nostrudel, for instance). Which is completely wrong. 

> You have indicated that your content should be read from a paid relay

Contents can be read from paid relays if the payment is only required to write into it, which is the case. Everybody else can read my posts, but only I can write. 

> So, 3 of 4 is where we should publish replies to you, and 3 of 4 you want us to read your events from

Correct.

> How does this give the network scalability or simplicity?

Because in theory, you only need 1 relay for your posts. You might need more relays for other stuff, like DMs, but for posts, it could only be 1. The rest is there for redundancy. This scales extremely well. Every person should have just 2-3 relays in there. And I recommend a mix between small, personal relays where people control, and 1 or 2 large ones as backup.

> It adds complexity for very little gained value.

I don't get where the complexity is in 3 relays. 

> My client will attempt to publish to all 4 of your relays.

That is almost correct. It should attempt to publish to 3 relays for read. And my client writes on 3 relays for write. 

> Users should use same relays for all their data,

Absolutely false. My DMs should never mix with my public data. Same for my health care data or my company's data (each has their own relays). My Drafts should only go to my private relay and none of the public ones or my DM relays. Having the same data everywhere is idiotic, wasteful, and creates lots of security and privacy issues. Calle's 14 relays proves the point.

> The mobile data usage is already out of hand, it won't get better continue down the similar path.

Sure, but the more relays you add, the worse it gets. You are giving up way too early on data usage.

NIP-65 should have just 2-3 relays. More than 1 just for redundancy in case one of them is too busy at a time. Every other relay type should go into their own lists.

**Comment by sondreb** [comment]
Call is not doing what I'm suggestion, I'm also not suggesting spreading your data on many relays. My suggestion is to have a small set of relays with all your data. Having few relays solves nothing by itself, when everyone is using the same 3-4 relays.

1. Discovery: Either from npub or nprofile, you should be able to find the relay of a user. What relay in general need to store, is only Relay List (kind:10002) for users, nothing more, for this discovery process to work and scale across millions of relays.

2. Profile: After discovering the relays, clients should go to those relays and get data on users, first the profile (kind:0) and then anything else that is interesting. It might be a good strategy to pick one random relay and only one, if that fails, fall back to a secondary one.

This removes the need for users to have many relays in their clients to discover data from other users. Everyone can have their own relays, and it works at global scale easily.

"Having the same data everywhere is idiotic, wasteful, and creates lots of security and privacy issues."

You have all your data on 2-4 relays, that's for backup reasons. There are no privacy issues, there are no security issues. It's not a problem to have your DMs on the same relay as your other data, as long as the relays don't give away your DM events to anyone other than you.

Having your public posts on X and your public post replies on Y and your DMs on Z, is overengineering and creates complexity where it gives no value, doesn't help scaling and it increases resource requirements for clients (an issue on mobile).

**Comment by vitorpamplona** [comment]
> You have all your data on 2-4 relays, that's for backup reasons. There are no privacy issues, there are no security issues.

??? I use over 20 relays these days, not 2-4. None of them are on my NIP-65 list as they should be. 

> There are no privacy issues, there are no security issues

I don't know where this is coming from. There are privacy issues everywhere. No relay implements all of the NIPs in just one relay. And likely never will. Even relay providers can't do it because the relay doesn't have enough information to do all of the roles in the first list, simply based on the connection data.

**Comment by vitorpamplona** [comment]
Let's not forget that we have NIPs like NIP-29 that specify a separate relay for each community/group of friends you have. Clients that implement NIP-29 should only use that list and ignore the NIP-65 or NIP-17 ones. Relays that implement NIP-29 can have completely different access restrictions. Otherwise, the data goes everywhere, and it is not supposed to.

Relays can be vastly different from one another. And there is the right and the wrong way to use each of them. So, I don't see any way of having just one small list for everything.

**Comment by sondreb** [comment]
> > You have all your data on 2-4 relays, that's for backup reasons. There are no privacy issues, there are no security issues.
> 
> ??? I use over 20 relays these days, not 2-4. None of them are on my NIP-65 list.
> 

You should (I don't want to demand anything, so let's say "could") only use those in your Relay List (kind:10002), nothing else ("for most use cases, there are many specialized use-cases that could be realized on Nostr) is needed.

It is not your job to add and connect to relays of another user, the Nostr client should do that for you, my client will jump to potentially thousands of different relays without problems. Just like I would jump around to different web servers if everyone had their personal blogs (like back in the days).

I'll continue implementing my suggestions and write more about them, and I'll share my findings and anyone who thinks there is any value in it, can consider adopting similar architecture.

**Comment by vitorpamplona** [comment]
> You should (I don't want to demand anything, so let's say "could") only use those in your Relay List (kind:10002), nothing else ("for most use cases, there are many specialized use-cases that could be realized on Nostr) is needed.

No, I should not. That would not only be against NIP-65 but would also have no usage for my followers or people messaging/replying to me (to which the `kind:10002` is designed for). 

> It is not your job to add and connect to relays of another user, the Nostr client should do that for you, my client will jump to potentially thousands of different relays without problems.

Sure, but your client will never figure out which relay is supposed to be used for which activity when joining me in some activities.

**Comment by mikedilger** [comment]
> my client will jump to potentially thousands of different relays without problems. Just like I would jump around to different web servers if everyone had their personal blogs (like back in the days).

That is the right way to think about it.

> Users should use same relays for all their data

This simply doesn't work.

Nostr is not a simple data storage system. And nostr is not a Twitter clone.  Nostr is an application-agnostic protocol enabling higher level applications.  Just like TCP/IP enables smtp, http, irc, etc, all running on top of it, nostr enables all kinds of creative things.  Many of those creative applications have crazy ways to use relays, very different from kind-10002 inbox/outbox stuff.  The inbox/outbox model works a treat for the Twitter-like part of nostr, but other applications will (and must) use different models depending on the details of what they are trying to do.

So, you are free to use the 10002 lists.  You are encouraged to only bother with the first 3 or 4 entries (and screw people who pack in lots of relays, make them suffer so they are encouraged to stop it).  And you can absolutely ignore all the other crazy ways that people choose to use relays, because they are at the application layer, not the nostr layer, and thus ignoring them does not break your usage of nostr in any way.  You can ignore kind 10050 NIP-17 DMs entirely if you want and it won't break anything.  Each application is opt-in, and separate from the base layer....   only the twitter-like base layer needs to be implemented by everybody.

Personally I want to see the nostr standard organized in this way, a base standard, and a bunch of application standards that it links to, instead of making it all like one giant glob.  But you can kinda figure out what is what when you read a NIP.

**Comment by mikedilger** [comment]
And so the main deal with this issue is that there are now lots of different rule sets by which relays operate.  And I can't just use any old relay for my outbox, or for my inbox, or for a chat server, or for DM receipt, or for an nsec bunker, or for an AI DVM service.... I have to know what the fuck this relay is configured to do, what rules is it operating by, so that I can make an informed choice when configuring my AI client app to use a relay that actually has an AI listening to it and responding to me, and not to some relay that is just an outbox service.  Right now clients (and users) are mostly still thinking of relays as generic and somewhat interchangeable, but they already are far from that.  So @vitorpamplona is trying (with this issue) to solve this problem so users know what they are doing.
